modules/bm_sim/src/lookup_structures.cpp

```diff
@@ -0,0 +1,252 @@
+    const auto entry_it = entries_map.find(key);
+    if (entry_it == entries_map.end()) {
+      return false;  // Nothing found
+    } else {
+      *handle = entry_it->second;
+      return true;
+    }
+  }
+
+  bool entry_exists(const ExactMatchKey &key) const override {
+    (void) key;
+    return entries_map.find(key.data) != entries_map.end();
+  }
+
+  void add_entry(const ExactMatchKey &key,
+                 internal_handle_t handle) override {
+    entries_map[key.data] = handle;  // key is copied, which is not great
+  }
+
+  void delete_entry(const ExactMatchKey &key) override {
+    entries_map.erase(key.data);
+  }
+
+  void clear() override {
+    entries_map.clear();
+  }
+
+ private:
+  std::unordered_map<ByteContainer, internal_handle_t, ByteContainerKeyHash>
+    entries_map{};
+};
+
+class TernaryMap : public TernaryLookupStructure {
+ public:
+  explicit TernaryMap(size_t nbytes_key)
+    : nbytes_key(nbytes_key) {}
+
+  bool lookup(const ByteContainer &key_data,
+              internal_handle_t *handle) const override {
+    int min_priority = std::numeric_limits<int>::max();;
+    bool match;
+
+    const TernaryMatchKey *entry;
+    const TernaryMatchKey *min_entry = nullptr;
+    entry_handle_t min_handle = 0;
+
+    for (auto it = this->handles.begin(); it != this->handles.end(); ++it) {
+      entry = &std::get<0>(*it);
+
+      if (entry->priority >= min_priority) continue;
+
+      match = true;
+      for (size_t byte_index = 0; byte_index < this->nbytes_key; byte_index++) {
+        if (entry->data[byte_index] !=
+            (key_data[byte_index] & entry->mask[byte_index])) {
+          match = false;
+          break;
+        }
+      }
+
+      if (match) {
+        min_priority = entry->priority;
+        min_entry = entry;
+        min_handle = std::get<1>(*it);
+      }
+    }
+
+    if (min_entry) {
+      *handle = min_handle;
+      return true;
+    }
+
+    return false;
+  }
+
+  bool entry_exists(const TernaryMatchKey &key) const override {
+    auto it = find_handle(key);
+
+    return it != this->handles.end();
+  }
+
+  void add_entry(const TernaryMatchKey &key,
+                 internal_handle_t handle) override {
+    // TODO(gordon) avoid copying key ?
+    handles.emplace_back(std::make_tuple(key, handle));
+  }
+
+  void delete_entry(const TernaryMatchKey &key) override {
+#if __GNUC__ < 4 || (__GNUC__ == 4 && (__GNUC_MINOR__ <= 8))
+    // In g++ 4.8 and lower vector::erase(const_iterator) is unimplemented
+    // this is a work-around for that shortcoming
+    // http://stackoverflow.com/questions/19559235/
+    //
+    auto c_it = find_handle(key);
+
+    // This advances a non-const iterator to the position pointed by the
+    // const iterator, effectively removing the const-ness of the iterator
+    auto it = this->handles.begin();
+    std::advance(it, std::distance(this->handles.cbegin(), c_it));
+#else
+    auto it = find_handle(key);
+#endif
```

Comment:
As the comment says, this is a workaround for an issue in `g++-4.8`. I've also tested it on 4.9 and it seems to work, but maybe it's worth also adding 4.9 to the travis build? As far as I can tell it's the default `g++` on the latest Debian (Jessie)